server/src/test/java/org/cloudfoundry/identity/uaa/oauth/jwt/JwtClientAuthenticationTest.java

```diff
@@ -129,21 +135,83 @@ void testGetClientAuthenticationParametersNullJwtClientConfiguration() {
+    assertTrue(params.containsKey("client_assertion"));
+    assertTrue(params.containsKey("client_assertion_type"));
+    JWSHeader header = validateClientAssertionOidcComplaint((String) params.get("client_assertion").get(0));
+    assertEquals("myKey", header.getKeyID());
+    assertNull(header.getJWKURL());
+  }
+
+  @Test
+  void testGetClientAssertionCustomSingingKeyButNoCertificate() throws ParseException {
+    // Given
+    mockKeyInfoService("myKey", null);
+    HashMap customClaims = new HashMap<>();
+    customClaims.put("kid", "myKey");
+    config.setJwtClientAuthentication(customClaims);
+    MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
+    // When
+    params = jwtClientAuthentication.getClientAuthenticationParameters(params, config);
+    // Then
+    assertTrue(params.containsKey("client_assertion"));
+    assertTrue(params.containsKey("client_assertion_type"));
+    validateClientAssertionOidcComplaint((String) params.get("client_assertion").get(0));
+    JWSHeader header = validateClientAssertionOidcComplaint((String) params.get("client_assertion").get(0));
+    assertEquals("myKey", header.getKeyID());
+    assertNotNull(header.getJWKURL());
+    assertEquals("http://localhost:8080/uaa/token_key", header.getJWKURL().toString());
+  }
+
+  private void mockKeyInfoService(String keyId, String x509Certificate) {
     KeyInfo keyInfo = mock(KeyInfo.class);
     Signer signer = mock(Signer.class);
+    if (keyId != null) {
+      KeyInfo customKeyInfo = mock(KeyInfo.class);
+      when(customKeyInfo.keyId()).thenReturn(keyId);
+      when(keyInfoService.getKey(keyId)).thenReturn(customKeyInfo);
+      when(customKeyInfo.algorithm()).thenReturn("RS256");
+      when(customKeyInfo.keyURL()).thenReturn("http://localhost:8080/uaa/token_key");
+      when(customKeyInfo.getSigner()).thenReturn(signer);
+      when(customKeyInfo.verifierCertificate()).thenReturn(x509Certificate != null ? Optional.of(x509Certificate): Optional.empty());
+    }
+    when(keyInfo.keyId()).thenReturn(KEY_ID);
+    when(keyInfoService.getKey(KEY_ID)).thenReturn(keyInfo);
     when(keyInfoService.getActiveKey()).thenReturn(keyInfo);
     when(keyInfo.algorithm()).thenReturn("RS256");
+    when(keyInfo.keyURL()).thenReturn("http://localhost:8080/uaa/token_key");
     when(keyInfo.getSigner()).thenReturn(signer);
-    when(keyInfo.verifierCertificate()).thenReturn(Optional.of(JwtHelperX5tTest.CERTIFICATE_1));
+    when(keyInfo.verifierCertificate()).thenReturn(x509Certificate != null ? Optional.of(x509Certificate): Optional.of(JwtHelperX5tTest.CERTIFICATE_1));
     when(signer.sign(any())).thenReturn("dummy".getBytes());
   }
 
-  private static void validateClientAssertionOidcComplaint(String clientAssertion) throws ParseException {
-    JWTClaimsSet jwtClaimsSet = JWTParser.parse(clientAssertion).getJWTClaimsSet();
+  private static JWSHeader validateClientAssertionOidcComplaint(String clientAssertion) throws ParseException {
```

Comment:
I think that the method name should be changed to represent the new implementation.

Reply:
Refactored this method can therefore stay as it is